README.md
# swagger2har

Transform you Swagger OAI spec files to a series of HAR request objects.

* http://swagger.io/specification/
* http://www.softwareishard.com/blog/har-12-spec/#request

## Setup

```bash
npm install
```

## Usage



Using as a ES module:
```js
import { swagger2har }  from 'swagger2har';
import swaggerJSON from 'your-swagger-api.json'; // e.g. http://petstore.swagger.io/v2/swagger.json

swagger2har(swaggerJSON);
```

To create HAR Request object for one path and method pair described in a given swagger use createHar directly

```js
import { createHar }  from 'swagger2har';
import swaggerJSON from 'your-swagger-api.json'; // e.g. http://petstore.swagger.io/v2/swagger.json

createHar(swaggerJSON, '/pet/findByTags', 'get', 'http://petstore.swagger.io');
```

## Testing

```bash
yarn test
```


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [1.0.6] - 2023-12-21

### Fixed

- Fixes a bug where `toUpperCase` is called on a value that doesn't exist, causing the response to break.

## [1.0.5] - 2023-03-06

### Fixed

- Fixes a bug with the example code snippets where an empty header value would break rendering if no type is present.

## [1.0.4] - 2023-01-30

### Fixed

- Fixes a bug with the example code snippets where an empty header value would break rendering.

## [1.0.3] - 2023-01-13

### Fixed

- Various package updates.

## [1.0.1] - 2021-05-24

### Fixed

- Support use of refs ( `"$ref"` ) in headers.

### Changed

- License set as Apache 2.0.

## [1.0.0] - 2021-05-21

### Added

- CHANGELOG.md introduced

### Changed

- Instead of default exporting swagger2har, we now `export { swagger2har, createHar }` This way apps can access createHar directly if needed.


__tests__/swagger-to-har.spec.js
import { swagger2har } from "../src"
import petStore from "./petstore.json"
import swagger2har_petstore_response from "./swagger2har_petstore_response.json"
import yaml from "js-yaml"
import fs from "fs"


test("swagger v2 JSON converts to HAR", () => {
  const HARs = swagger2har(petStore)
  const add_a_new_pet = HARs[0]

  expect(JSON.stringify(swagger2har_petstore_response)).toEqual(JSON.stringify(HARs))
})

test("swagger v2 JSON to HAR object can be crawled", () => {
  const HARs = swagger2har(petStore)
  const add_a_new_pet = HARs[0]
  const postObj = JSON.parse(add_a_new_pet.har.postData.text)

  expect(add_a_new_pet.har.method).toEqual("POST")
  expect(add_a_new_pet.har.url).toEqual("http://petstore.swagger.io/v2/pet")
  expect(add_a_new_pet.har.httpVersion).toEqual("HTTP/1.1")
  expect(postObj).toHaveProperty("category.name")
})

test("swagger v2 YAML to JSON converts to HAR", () => {
  const vitals = yaml.load(fs.readFileSync(process.cwd() + "/__tests__/vitals.yaml", "utf8"))
  const HARs = swagger2har(vitals)
  const get_cluster_stats = HARs[0]

  expect(get_cluster_stats.har.method).toEqual("GET")
  expect(get_cluster_stats.har.url).toEqual("http://127.0.0.1:8001/vitals/")
  expect(get_cluster_stats.har.httpVersion).toEqual("HTTP/1.1")
})

test("swagger v3 YAML to JSON converts to HAR", () => {
  const petstore = yaml.load(fs.readFileSync(process.cwd() + "/__tests__/petstore-expanded-v3.yaml", "utf8"))
  const HARs = swagger2har(petstore)
  const pets = HARs[0]

  expect(pets.har.method).toEqual("GET")
  expect(pets.har.url).toEqual("http://petstore.swagger.io/api/pets")
  expect(pets.har.httpVersion).toEqual("HTTP/1.1")
})

test("swagger v3 JSON with header params converts to HAR", () => {
  const spec = yaml.load(fs.readFileSync(process.cwd() + "/__tests__/header-params.json", "utf8"))
  const HARs = swagger2har(spec)
  const res = HARs[0]

  expect(res.har.method).toEqual("GET")
  expect(res.har.url).toEqual("http://localhost/status/{statusId}")
  expect(res.har.httpVersion).toEqual("HTTP/1.1")

  const headers = res.har.headers

  expect(headers[0].name).toEqual("accept-encoding")
  expect(headers[0].value).toEqual("<SOME_STRING_VALUE>")

  expect(headers[1].name).toEqual("user-agent")
  expect(headers[1].value).toEqual("<SOME_STRING_VALUE>")

  expect(headers[2].name).toEqual("connection")
  expect(headers[2].value).toEqual("<SOME_STRING_VALUE>")
})

test("swagger v3 JSON with $ref header params converts to HAR", () => {
  const spec = yaml.load(fs.readFileSync(process.cwd() + "/__tests__/header-params.json", "utf8"))
  const HARs = swagger2har(spec)
  const res = HARs[1]

  expect(res.har.method).toEqual("GET")
  expect(res.har.url).toEqual("http://localhost/ref-header")
  expect(res.har.httpVersion).toEqual("HTTP/1.1")

  const headers = res.har.headers

  expect(headers[0].name).toEqual("special-x")
  expect(headers[0].value).toEqual("<SOME_STRING_VALUE>")
})


src/converter.js
/**
 * Translates given Swagger 2.0 file to an array of HTTP Archive (HAR) 1.2 Request Object.
 * See more:
 *  - http://swagger.io/specification/
 *  - http://www.softwareishard.com/blog/har-12-spec/#request
 *
 * Example HAR Request Object:
 * "request": {
 *   "method": "GET",
 *   "url": "http://www.example.com/path/?param=value",
 *   "httpVersion": "HTTP/1.1",
 *   "cookies": [],
 *   "headers": [],
 *   "queryString" : [],
 *   "postData" : {},
 *   "headersSize" : 150,
 *   "bodySize" : 0,
 *   "comment" : ""
 * }
 *
 * Source code initially pulled from: https://github.com/ErikWittern/swagger-snippet/blob/master/swagger-to-har.js
 */

var OpenAPISampler = require("openapi-sampler")

/**
 * Create HAR Request object for path and method pair described in given swagger.
 *
 * @param  {Object} swagger           Swagger document
 * @param  {string} path              Key of the path
 * @param  {string} method            Key of the method
 * @param  {Object} queryParamValues  Optional: Values for the query parameters if present
 * @return {Object}                   HAR Request object
 */
var createHar = function(swagger, path, method, queryParamValues) {
  // if the operational parameter is not provided, set it to empty object
  if (typeof queryParamValues === "undefined") {
    queryParamValues = {}
  }

  var baseUrl = getBaseUrl(swagger)

  var har = {
    method: method.toUpperCase(),
    url: baseUrl + path,
    headers: getHeadersArray(swagger, path, method),
    queryString: getQueryStrings(swagger, path, method, queryParamValues),
    httpVersion: "HTTP/1.1",
    cookies: [],
    headersSize: 0,
    bodySize: 0
  }

  // get payload data, if available:
  var postData = getPayload(swagger, path, method)
  if (postData) har.postData = postData

  return har
}

/**
 * Get the payload definition for the given endpoint (path + method) from the
 * given OAI specification. References within the payload definition are
 * resolved.
 *
 * @param  {object} swagger
 * @param  {string} path
 * @param  {string} method
 * @return {object}
 */
var getPayload = function(swagger, path, method) {
  if (typeof swagger.paths[path][method].parameters !== "undefined") {
    for (var i in swagger.paths[path][method].parameters) {
      var param = swagger.paths[path][method].parameters[i]
      if (typeof param.in !== "undefined" && param.in.toLowerCase() === "body" &&
        typeof param.schema !== "undefined") {
        try {
          const sample = OpenAPISampler.sample(param.schema, { skipReadOnly: true }, swagger)
          return {
            mimeType: "application/json",
            text: JSON.stringify(sample)
          }
        } catch (err) {
          console.log(err)
          return null
        }
      }
    }
  }
  if (swagger.paths[path][method].requestBody && swagger.paths[path][method].requestBody.content &&
    swagger.paths[path][method].requestBody.content["application/json"] &&
    swagger.paths[path][method].requestBody.content["application/json"].schema) {
    const sample = OpenAPISampler.sample(swagger.paths[path][method].requestBody.content["application/json"].schema, { skipReadOnly: true }, swagger)
    return {
      mimeType: "application/json",
      text: JSON.stringify(sample)
    }
  }
  return null
}

/**
 * Gets the base URL constructed from the given swagger.
 *
 * @param  {Object} swagger Swagger document
 * @return {string}         Base URL
 */
var getBaseUrl = function(swagger) {
  if (swagger.servers)
    return swagger.servers[0].url

  var baseUrl = ""

  if (typeof swagger.schemes !== "undefined") {
    baseUrl += swagger.schemes[0]
  } else {
    baseUrl += "http"
  }

  baseUrl += "://" + swagger.host

  if (typeof swagger.basePath !== "undefined" &&
    swagger.basePath !== "/") {
    baseUrl += swagger.basePath
  }

  return baseUrl
}

/**
 * Get array of objects describing the query parameters for a path and method pair
 * described in the given swagger.
 *
 * @param  {Object} swagger Swagger document
 * @param  {string} path    Key of the path
 * @param  {string} method  Key of the method
 * @param  {Object} values  Optional: query parameter values to use in the snippet if present
 * @return {array}          List of objects describing the query strings
 */
var getQueryStrings = function(swagger, path, method, values) {
  // Set the optional parameter if it's not provided
  if (typeof values === "undefined") {
    values = {}
  }

  var queryStrings = []

  if (typeof swagger.paths[path][method].parameters !== "undefined") {
    for (var i in swagger.paths[path][method].parameters) {
      var param = swagger.paths[path][method].parameters[i]
      if (typeof param["$ref"] === "string" &&
        !/^http/.test(param["$ref"])) {
        param = resolveRef(swagger, param["$ref"])
      }
      if (typeof param.in !== "undefined" && param.in.toLowerCase() === "query") {
        queryStrings.push({
          name: param.name,
          value: typeof values[param.name] === "undefined"
            ? (typeof param.default === "undefined"
              ? ("SOME_" + (param.type || param.schema.type).toUpperCase() + "_VALUE")
              : param.default + "")
            : (values[param.name] + "") /* adding a empty string to convert to string */
        })
      }
    }
  }

  return queryStrings
}

/**
 * Get an array of objects describing the header for a path and method pair
 * described in the given swagger.
 *
 * @param  {Object} swagger Swagger document
 * @param  {string} path    Key of the path
 * @param  {string} method  Key of the method
 * @return {array}          List of objects describing the header
 */
var getHeadersArray = function(swagger, path, method) {
  var headers = []

  var pathObj = swagger.paths[path][method]

  // 'accept' header:
  if (typeof pathObj.consumes !== "undefined") {
    for (var i in pathObj.consumes) {
      var type = pathObj.consumes[i]
      headers.push({
        name: "accept",
        value: type
      })
    }
  }

  // 'content-type' header:
  if (typeof pathObj.produces !== "undefined") {
    for (var j in pathObj.produces) {
      var type2 = pathObj.produces[j]
      headers.push({
        name: "content-type",
        value: type2
      })
    }
  }

  // v3 'content-type' header:
  if (pathObj.requestBody && pathObj.requestBody.content) {
    for (const type3 of Object.keys(pathObj.requestBody.content)) {
      headers.push({
        name: "content-type",
        value: type3
      })
    }
  }

  // headers defined in path object:
  if (typeof pathObj.parameters !== "undefined") {
    for (var k in pathObj.parameters) {
      var param = pathObj.parameters[k]
      if (typeof param.in !== "undefined" && param.in.toLowerCase() === "header") {
        headers.push({
          name: param.name,
          value: "SOME_" + (param.type || param.schema.type).toUpperCase() + "_VALUE"
        })
      }
    }
  }

  // security:
  var basicAuthDef
  var apiKeyAuthDef
  var oauthDef
  if (typeof pathObj.security !== "undefined") {
    for (var l in pathObj.security) {
      var secScheme = Object.keys(pathObj.security[l])[0]
      var secDefinition = swagger.securityDefinitions ?
        swagger.securityDefinitions[secScheme] :
        swagger.components.securitySchemes[secScheme]
      var authType = secDefinition.type.toLowerCase()
      switch (authType) {
        case "basic":
          basicAuthDef = secScheme
          break
        case "apikey":
          if (secDefinition.in === "header") {
            apiKeyAuthDef = secDefinition
          }
          break
        case "oauth2":
          oauthDef = secScheme
          break
      }
    }
  } else if (typeof swagger.security !== "undefined") {
    // Need to check OAS 3.0 spec about type http and scheme
    for (var m in swagger.security) {
      var secScheme = Object.keys(swagger.security[m])[0]
      var secDefinition = swagger.components.securitySchemes[secScheme]
      var authType = secDefinition.type.toLowerCase()
      let authScheme = secDefinition.scheme.toLowerCase()
      switch (authType) {
        case "http":
          switch (authScheme) {
            case "bearer":
              oauthDef = secScheme
              break
            case "basic":
              basicAuthDef = secScheme
              break
          }
          break
        case "basic":
          basicAuthDef = secScheme
          break
        case "apikey":
          if (secDefinition.in === "header") {
            apiKeyAuthDef = secDefinition
          }
          break
        case "oauth2":
          oauthDef = secScheme
          break
      }
    }
  }

  if (basicAuthDef) {
    headers.push({
      name: "Authorization",
      value: "Basic " + "REPLACE_BASIC_AUTH"
    })
  } else if (apiKeyAuthDef) {
    headers.push({
      name: apiKeyAuthDef.name,
      value: "REPLACE_KEY_VALUE"
    })
  } else if (oauthDef) {
    headers.push({
      name: "Authorization",
      value: "Bearer " + "REPLACE_BEARER_TOKEN"
    })
  }

  return headers
}

/**
 * Produces array of HAR files for given Swagger document
 *
 * @param  {object}   swagger          A swagger document
 * @param  {Function} callback
 */
var swaggerToHarList = function(swagger) {
  try {
    // determine basePath:
    var baseUrl = getBaseUrl(swagger)

    // iterate Swagger and create har objects:
    var harList = []
    for (var path in swagger.paths) {
      for (var method in swagger.paths[path]) {
        var url = baseUrl + path
        var har = createHar(swagger, path, method)
        harList.push({
          method: method.toUpperCase(),
          url: url,
          description: swagger.paths[path][method].description || "No description available",
          har: har
        })
      }
    }

    return harList
  } catch (e) {
    console.log(e)
  }
}

/**
 * Returns the value referenced in the given reference string
 *
 * @param  {object} oai
 * @param  {string} ref A reference string
 * @return {any}
 */
var resolveRef = function(oai, ref) {
  var parts = ref.split("/")

  if (parts.length <= 1) return {} // = 3

  var recursive = function(obj, index) {
    if (index + 1 < parts.length) { // index = 1
      var newCount = index + 1
      return recursive(obj[parts[index]], newCount)
    } else {
      return obj[parts[index]]
    }
  }
  return recursive(oai, 1)
}

module.exports = {
  getAll: swaggerToHarList,
  getEndpoint: createHar
}


src/swagger2har.js
/**
 * Translates given Swagger 2.0 file to an array of HTTP Archive (HAR) 1.2 Request Object.
 * See more:
 *  - http://swagger.io/specification/
 *  - http://www.softwareishard.com/blog/har-12-spec/#request
 *
 * Example HAR Request Object:
 * "request": {
 *   "method": "GET",
 *   "url": "http://www.example.com/path/?param=value",
 *   "httpVersion": "HTTP/1.1",
 *   "cookies": [],
 *   "headers": [],
 *   "queryString" : [],
 *   "postData" : {},
 *   "headersSize" : 150,
 *   "bodySize" : 0,
 *   "comment" : ""
 * }
 *
 * Source code initially pulled from: https://github.com/ErikWittern/swagger-snippet/blob/master/swagger-to-har.js
 */

import instantiator from "json-schema-instantiator"

/**
 * Create HAR Request object for path and method pair described in given swagger.
 *
 * @param  {Object} swagger           Swagger document
 * @param  {string} path              Key of the path
 * @param  {string} method            Key of the method
 * @param  {string} baseUrl           Base URL
 * @param  {string} operation         Key of the operation
 * @param  {Object} queryParamValues  Optional: Values for the query parameters if present
 * @return {Object}                   HAR Request object
 */

var createHar = function(swagger, path, method, baseUrl, queryParamValues) {
  // if the operational parameter is not provided, set it to empty object
  if (typeof queryParamValues === "undefined") {
    queryParamValues = {}
  }

  var har = {
    method: method.toUpperCase(),
    url: baseUrl + path,
    headers: getHeadersArray(swagger, path, method),
    queryString: getQueryStrings(swagger, path, method, queryParamValues),
    httpVersion: "HTTP/1.1",
    cookies: [],
    headersSize: 0,
    bodySize: 0
  }

  // get payload data, if available:
  var postData = getPayload(swagger, path, method)
  if (postData) har.postData = postData

  return har
}

/**
 * Get the payload definition for the given endpoint (path + method) from the
 * given OAI specification. References within the payload definition are
 * resolved.
 *
 * @param  {object} swagger
 * @param  {string} path
 * @param  {string} method
 * @return {object}
 */
var getPayload = function(swagger, path, method) {
  if (typeof swagger.paths[path][method].parameters !== "undefined") {
    for (var i in swagger.paths[path][method].parameters) {
      var param = swagger.paths[path][method].parameters[i]
      if (
        typeof param.in !== "undefined" &&
        param.in.toLowerCase() === "body" &&
        typeof param.schema !== "undefined"
      ) {
        var schema
        if (typeof param.schema["$ref"] === "undefined") {
          schema = param.schema
        } else if (/^http/.test(param.schema["$ref"])) {
          // can't resolve this for now...
        } else {
          var ref = param.schema["$ref"].split("/").slice(-1)[0]
          schema = getResolvedSchema(swagger, swagger.definitions[ref])
        }

        return {
          mimeType: "application/json",
          text: JSON.stringify(instantiator.instantiate(schema))
        }
      }
    }
  }
  return null
}

/**
 * Get a complete JSON schema from Swagger, where all references ($ref) are
 * resolved. $ref appear:
 * - as properties
 * - as items
 *
 * @param  {[type]} swagger [description]
 * @param  {[type]} schema  [description]
 * @param  {[type]} ref     [description]
 * @return {[type]}         [description]
 */
var getResolvedSchema = function(swagger, schema) {
  if (schema.type === "object") {
    if (typeof schema.properties !== "undefined") {
      for (var propKey in schema.properties) {
        var prop = schema.properties[propKey]
        if (typeof prop["$ref"] === "string" && !/^http/.test(prop["$ref"])) {
          var ref = prop["$ref"].split("/").slice(-1)[0]
          schema.properties[propKey] = swagger.definitions[ref]
        }
        getResolvedSchema(swagger, schema.properties[propKey])
      }
    }
  } else if (schema.type === "array") {
    if (typeof schema.items !== "undefined") {
      for (var itemKey in schema.items) {
        if (itemKey === "$ref" && !/^http/.test(schema.items[itemKey])) {
          var ref2 = schema.items["$ref"].split("/").slice(-1)[0]
          schema.items = swagger.definitions[ref2]
        }
        getResolvedSchema(swagger, schema.items)
      }
    }
  }
  return schema
}

/**
 * Gets the base URL constructed from the given swagger.
 *
 * @param  {Object} swagger Swagger document
 * @return {string}         Base URL
 */
var getBaseUrl = function(swagger) {
  var baseUrl = ""

  if (swagger.openapi) {
    return swagger.servers[0].url
  }

  if (typeof swagger.schemes !== "undefined") {
    baseUrl += swagger.schemes[0]
  } else {
    baseUrl += "http"
  }

  if (swagger.basePath === "/") {
    baseUrl += "://" + swagger.host
  } else {
    baseUrl += "://" + swagger.host + swagger.basePath
  }

  return baseUrl
}

/**
 * Get array of objects describing the query parameters for a path and method pair
 * described in the given swagger.
 *
 * @param  {Object} swagger Swagger document
 * @param  {string} path    Key of the path
 * @param  {string} method  Key of the method
 * @param  {Object} values  Optional: query parameter values to use in the snippet if present
 * @return {array}          List of objects describing the query strings
 */
var getQueryStrings = function(swagger, path, method, values) {
  // Set the optional parameter if it's not provided
  if (typeof values === "undefined") {
    values = {}
  }

  var queryStrings = []

  if (typeof swagger.paths[path][method].parameters !== "undefined") {
    for (var i in swagger.paths[path][method].parameters) {
      var param = swagger.paths[path][method].parameters[i]
      if (typeof param["$ref"] === "string" && !/^http/.test(param["$ref"])) {
        param = resolveRef(swagger, param["$ref"])
      }
      if (typeof param.in !== "undefined" && param.in.toLowerCase() === "query") {
        queryStrings.push({
          name: param.name,
          value:
            typeof values[param.name] === "undefined"
              ? typeof param.default === "undefined"
                ? swagger.openapi
                  ? param.schema?.type ? "<SOME_" + param.schema?.type?.toUpperCase() + "_VALUE>" : "<SOME_VALUE>"
                  : param.type ? "<SOME_" + param.type?.toUpperCase() + "_VALUE>" : "<SOME_VALUE>"
                : param.default + ""
              : values[param.name] + "" /* adding a empty string to convert to string */
        })
      }
    }
  }

  return queryStrings
}

/**
 * Get an array of objects describing the header for a path and method pair
 * described in the given swagger.
 *
 * @param  {Object} swagger Swagger document
 * @param  {string} path    Key of the path
 * @param  {string} method  Key of the method
 * @return {array}          List of objects describing the header
 */
var getHeadersArray = function(swagger, path, method) {
  var headers = []

  var pathObj = swagger.paths[path][method]

  // 'accept' header:
  if (typeof pathObj.consumes !== "undefined") {
    for (var i in pathObj.consumes) {
      var type = pathObj.consumes[i]
      headers.push({
        name: "accept",
        value: type
      })
    }
  }

  // 'content-type' header:
  if (typeof pathObj.produces !== "undefined") {
    for (var j in pathObj.produces) {
      var type2 = pathObj.produces[j]
      headers.push({
        name: "content-type",
        value: type2
      })
    }
  }

  // headers defined in path object:
  if (typeof pathObj.parameters !== "undefined") {
    for (var k in pathObj.parameters) {
      var param = pathObj.parameters[k]
      if (typeof param.in !== "undefined" && param.in.toLowerCase() === "header") {
        if (typeof param["$ref"] === "string") {
          // can't handle this yet
          if (/^http/.test(param["$ref"])) {
            continue
          }
          param = resolveRef(swagger, param["$ref"])
        }
        var paramType = swagger.openapi ? param.schema.type: param.type
        headers.push({
          name: param.name,
          value: paramType ? `<SOME_${paramType?.toUpperCase()}_VALUE>` : "<SOME_VALUE>"
        })
      }
    }
  }

  // security:
  var basicAuthDef
  var apiKeyAuthDef
  var oauthDef

  var secDefs = swagger.securityDefinitions || swagger.components && swagger.components.securitySchemes
  var securityObj = pathObj.security || swagger.security

  if (secDefs && securityObj) {
    for (var l in securityObj) {
      var secScheme = Object.keys(securityObj[l])[0]

      var def = secDefs[secScheme]
      if (!def || !def.type) {
        continue
      }

      var authType = def.type.toLowerCase()
      switch (authType) {
        case "basic":
          basicAuthDef = secScheme
          break
        case "apikey":
          if (def.in === "query") {
            apiKeyAuthDef = secScheme
          }
          break
        case "oauth2":
          oauthDef = secScheme
          break
      }
    }
  }

  if (basicAuthDef) {
    headers.push({
      name: "Authorization",
      value: "Basic " + "<USERNAME:PASSWORD>"
    })
  } else if (apiKeyAuthDef) {
    headers.push({
      name: secDefs[apiKeyAuthDef].name,
      value: "REPLACE_KEY_VALUE"
    })
  } else if (oauthDef) {
    headers.push({
      name: "Authorization",
      value: "Bearer " + "<BEARER_TOKEN>"
    })
  }

  return headers
}

/**
 * Produces array of HAR files for given Swagger document
 *
 * @param  {object} swagger A swagger JSON document
 * @param  {String} [selectedServer] Optional selected server to use for har
 * @param  {Function} callback
 */
var swagger2har = function(swagger, selectedServer) {

  try {
    // determine basePath:
    var baseUrl = selectedServer || getBaseUrl(swagger)

    var harList = []
    // iterate over Swagger paths and create HAR objects
    Object.keys(swagger.paths).forEach(path => {
      Object.keys(swagger.paths[path]).forEach(operation => {
        var url = baseUrl + path
        var har = createHar(swagger, path, operation, baseUrl)
        harList.push({
          path,
          method: operation?.toUpperCase(),
          url: url,
          description: swagger.paths[path][operation].description || "No description available",
          har: har
        })
      })
    })

    return harList
  } catch (e) {
    console.error(e)
    return null
  }
}

/**
 * Returns the value referenced in the given reference string
 *
 * @param  {object} oai
 * @param  {string} ref A reference string
 * @return {any}
 */
var resolveRef = function(oai, ref) {
  var parts = ref.split("/")

  if (parts.length <= 1) return {} // = 3

  var recursive = function(obj, index) {
    if (index + 1 < parts.length) {
      // index = 1
      var newCount = index + 1
      return recursive(obj[parts[index]], newCount)
    } else {
      return obj[parts[index]]
    }
  }
  return recursive(oai, 1)
}

export { swagger2har, createHar }
